experimental/metal/metal_device.m

```diff
@@ -0,0 +1,522 @@
+  return iree_hal_metal_driver_device_params(device->driver);
+}
+
+id<MTLCommandQueue> iree_hal_metal_device_command_queue(const iree_hal_device_t* base_device) {
+  const iree_hal_metal_device_t* device = iree_hal_metal_device_const_cast(base_device);
+  return device->queue;
+}
+
+static iree_status_t iree_hal_metal_device_create_internal(
+    iree_hal_driver_t* driver, iree_string_view_t identifier,
+    const iree_hal_metal_device_params_t* params, id<MTLDevice> metal_device,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device) {
+  iree_hal_metal_device_t* device = NULL;
+
+  iree_host_size_t total_size = iree_sizeof_struct(*device) + identifier.size;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(host_allocator, total_size, (void**)&device));
+  memset(device, 0, total_size);
+
+  iree_status_t status = iree_hal_metal_allocator_create((iree_hal_device_t*)device, metal_device,
+                                                         params->resource_hazard_tracking_mode,
+                                                         host_allocator, &device->device_allocator);
+
+  iree_hal_metal_builtin_executable_t* builtin_executable = NULL;
+  if (iree_status_is_ok(status)) {
+    status =
+        iree_hal_metal_builtin_executable_create(metal_device, host_allocator, &builtin_executable);
+  } else {
+    iree_hal_device_release((iree_hal_device_t*)device);
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_metal_staging_buffer_initialize(
+        metal_device, params->queue_uniform_buffer_size, &device->staging_buffer);
+  } else {
+    iree_hal_device_release((iree_hal_device_t*)device);
+  }
+
+  if (iree_status_is_ok(status)) {
+    iree_hal_resource_initialize(&iree_hal_metal_device_vtable, &device->resource);
+    iree_string_view_append_to_buffer(identifier, &device->identifier,
+                                      (char*)device + iree_sizeof_struct(*device));
+    iree_arena_block_pool_initialize(params->arena_block_size, host_allocator, &device->block_pool);
+    device->driver = driver;
+    iree_hal_driver_retain(device->driver);
+    device->host_allocator = host_allocator;
+    device->device = [metal_device retain];          // +1
+    device->queue = [metal_device newCommandQueue];  // +1
+    device->command_buffer_resource_reference_mode = params->command_buffer_resource_reference_mode;
+    device->builtin_executable = builtin_executable;
+    dispatch_queue_attr_t queue_attr = dispatch_queue_attr_make_with_qos_class(
+        DISPATCH_QUEUE_SERIAL, QOS_CLASS_USER_INITIATED, /*relative_priority=*/0);
+    device->semaphore_notification_queue =
+        dispatch_queue_create("dev.iree.queue.metal", queue_attr);
+    device->event_listener = [[MTLSharedEventListener alloc]
+        initWithDispatchQueue:device->semaphore_notification_queue];  // +1
+    device->capture_manager = NULL;
+    *out_device = (iree_hal_device_t*)device;
+  }
+  return iree_ok_status();
```

Comment:
Return `status`, not `iree_ok_status()` (also maybe audit the rest of the code for that error)
```suggestion
  return status;
```

Reply:
Good catch!